---
title: "annotables"
output: github_document
---

```{r setup, include=FALSE}
library(knitr)
opts_chunk$set(
    cache = TRUE,
    echo = TRUE,
    eval = FALSE,
    message = FALSE,
    warning = FALSE)
```

Tables for annotating gene lists and converting between identifiers.

## Why?

Many bioinformatics tasks require converting gene identifiers from one convention to another, or annotating gene identifiers with gene symbol, description, position, etc. Sure, [biomaRt](https://bioconductor.org/packages/release/bioc/html/biomaRt.html) does this for you, but I got tired of remembering biomaRt syntax and hammering Ensembl's servers every time I needed to do this.

This package has basic annotation information from [Ensembl release 89](http://may2017.archive.ensembl.org/) for:

- Human build 38 (`grch38`)
- Human build 37 (`grch37`)
- Mouse (`grcm38`)
- Rat (`rnor6`)
- Chicken (`galgal5`)
- Worm (`wbcel235`)
- Fly (`bdgp6`)

Where each table contains:

- `ensgene`: Ensembl gene ID
- `entrez`: Entrez gene ID
- `symbol`: Gene symbol
- `chr`: Chromosome
- `start`: Start
- `end`: End
- `strand`: Strand
- `biotype`: Protein coding, pseudogene, mitochondrial tRNA, etc.
- `description`: Full gene name/description.

Additionally, there are tables for human and mouse (`grch38_gt`, `grch37_gt`, and `grcm38_gt`, respectively) that link ensembl gene IDs to ensembl transcript IDs.

## Usage

Installation:

```{r}
install.packages("devtools")
devtools::install_github("stephenturner/annotables")
```

It isn't necessary to load dplyr, but the tables are `tibble` and will print nicely if you have dplyr loaded.

```{r, eval=TRUE}
library(dplyr)
library(annotables)
```

Look at the human genes table (note the description column gets cut off because the table becomes too wide to print nicely):

```{r, eval=TRUE}
grch38
```

Look at the human genes-to-transcripts table:

```{r, eval=TRUE}
grch38_tx2gene
```

Tables are `tibble`, pipe-able with dplyr:

```{r, eval=TRUE, results='asis'}
grch38 %>% 
    filter(biotype == "protein_coding" & chr == "1") %>% 
    select(ensgene, symbol, chr, start, end, description) %>% 
    head %>% 
    kable
```


Example with DESeq2 results from the [airway](https://bioconductor.org/packages/release/data/experiment/html/airway.html) package, made tidy with [biobroom](http://www.bioconductor.org/packages/devel/bioc/html/biobroom.html):

```{r, eval=TRUE}
library(DESeq2)
library(airway)

data(airway)
airway <- DESeqDataSet(airway, design = ~cell + dex)
airway <- DESeq(airway)
res <- results(airway)

# tidy results with biobroom
library(biobroom)
res_tidy <- tidy.DESeqResults(res)
head(res_tidy)
```

```{r, eval=TRUE, results='asis'}
res_tidy %>% 
    arrange(p.adjusted) %>% 
    head(20) %>% 
    inner_join(grch38, by = c("gene" = "ensgene")) %>% 
    select(gene, estimate, p.adjusted, symbol, description) %>% 
    kable
```
